<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span><ul class="toc-item"><li><span><a href="#Информация-о-данных" data-toc-modified-id="Информация-о-данных-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Информация о данных</a></span></li></ul></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span><ul class="toc-item"><li><span><a href="#Выделение-признаков." data-toc-modified-id="Выделение-признаков.-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Выделение признаков.</a></span></li><li><span><a href="#Создание-модели-линейной-регрессии" data-toc-modified-id="Создание-модели-линейной-регрессии-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Создание модели линейной регрессии</a></span></li><li><span><a href="#Создание-обратимой-матрицы" data-toc-modified-id="Создание-обратимой-матрицы-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Создание обратимой матрицы</a></span></li><li><span><a href="#Обучение-модели-на-исходных-данных" data-toc-modified-id="Обучение-модели-на-исходных-данных-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Обучение модели на исходных данных</a></span></li><li><span><a href="#Обучение-модели-на-преобразованных-данных" data-toc-modified-id="Обучение-модели-на-преобразованных-данных-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Обучение модели на преобразованных данных</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li><li><span><a href="#Финальные-комментарии-ревьюера" data-toc-modified-id="Финальные-комментарии-ревьюера-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Финальные комментарии ревьюера</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

Импортируем библиотеки, которые могут понадобиться в ходе написания проекта.

In [1]:
import pandas as pd
import numpy as np
from sklearn.metrics import r2_score

### Информация о данных

Прочитаем файл, сохраним его в переменной data и просмотрим его.

In [2]:
data = pd.read_csv('/datasets/insurance.csv')

Напишем функцию, которая принимает датасет и возвращает различную информацию о данных в нем.

In [3]:
def data_info (df):
    display(df.head())
    print('Информация о датасете:')
    df.info()
    print('Описательная статистика:')
    display(df.describe().T)
    print('Дубликаты:')
    display(df.duplicated().sum())
    print('Пропуски:')
    for col in df.columns:
        pct_missing = np.mean(df[col].isna())
        print('{} - {}%'.format(col, round(pct_missing*100,2)))

In [4]:
data_info(data)

Пол  Возраст  Зарплата  Члены семьи  Страховые выплаты
0    1     41.0   49600.0            1                  0
1    0     46.0   38000.0            1                  1
2    0     29.0   21000.0            0                  0
3    0     21.0   41700.0            2                  0
4    1     28.0   26100.0            0                  0

Информация о датасете:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB
Описательная статистика:


count        mean          std     min      25%      50%  \
Пол                5000.0      0.4990     0.500049     0.0      0.0      0.0   
Возраст            5000.0     30.9528     8.440807    18.0     24.0     30.0   
Зарплата           5000.0  39916.3600  9900.083569  5300.0  33300.0  40200.0   
Члены семьи        5000.0      1.1942     1.091387     0.0      0.0      1.0   
Страховые выплаты  5000.0      0.1480     0.463183     0.0      0.0      0.0   

                       75%      max  
Пол                    1.0      1.0  
Возраст               37.0     65.0  
Зарплата           46600.0  79000.0  
Члены семьи            2.0      6.0  
Страховые выплаты      0.0      5.0

Дубликаты:


153

Пропуски:
Пол - 0.0%
Возраст - 0.0%
Зарплата - 0.0%
Члены семьи - 0.0%
Страховые выплаты - 0.0%


Дата сет состоит из 5 столбцов и 5000 строк.
- Столбец пол имеет тип данных int - целочисленные значения 1 или 0. В описании к данным не указано какой пол соответствут значениям 1 или 0, это не так важно. Предположим что 0 - женский пол, а 1 - мужской.
- Столбец Возраст имеет тип данных float. Значения возраста. Минимальное значение 18 лет максимальное 65.
- В столбце Зарплата указаны данные о зарплате клиентов. Минимальное значение 5300, а максимальное 79000. Средняя зарплата 39916, а медианное значение 40200. В описании к данным не указано в каких единицах измеряется зарплата, предположим что в рублях.
- Столбец члены семьи содержит данные о количестве членов в семье и имеет тип данных int. 
- Страховые выплаты. Тип данных int, от 0 до 5, это говорит о том что в данном столбце представленны данные о количестве случаев страховых выплат, а не о их размере.<br><br>
В данных отсутствуют пропуски, но имеются дубликаты 153 случая. Значения в данных по своей природе не сильно отличаются друг от друга и дубликаты вполне могут являться не дубликатами, а данными разных клиентов, поэтому, я считаю удалять их не совсем корректно.

## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

**Обоснование:** 
Умножим признаки на обратимую матрицу.<br><br>
Если:<br>
$$Хn=XP$$<br><br>
Тогда:<br>
$$w_{n} = ((XP)^T XP)^{-1} (XP)^T y$$

$$w_{n} = ((P^TX^T)(XP))^{-1} (P^TX^T) y$$    
    
$$w_{n} = (P^T((X^TX)P))^{-1} (P^TX^T) y$$

$$w_{n} = ((X^TX)P)^{-1}(P^T)^{-1}(P^TX^T)y$$

$$w_{n} = P^{-1}(X^TX)^{-1}((P^T)^{-1}(P^T))X^T y$$
    
$$w_{n} = P^{-1}(X^TX)^{-1}X^T y$$<br><br>
Имея формулу обучения:
$$
w = (X^T X)^{-1} X^T y
$$
Получаем:
$$
w_{n} = P^{-1}w
$$
<br><br>
Далее подставим наши значения в формулу предсказания:
$$a = Xw$$
Наша формула предсказания будет иметь вид:
$$a_{n} = X_{n}w_{n}$$

А если: $$X_{n} = XP$$
и $$w_{n} = P^{-1}w$$

Получается что:
$$a_{n} = XP P^{-1}w$$
$$a_{n} = Xw$$

Исходя из этого видно что:
$$a_{n} = Xw = a$$

**Ответ:** Верный ответ  b. не изменится. $a_{n}$ это тоже самое что $a$, а значит при умножении матрицы признаков на любую обратимую матрицу качество обучения модели линейной регрессии не меняется.

## Алгоритм преобразования

**Алгоритм**

Для того чтобы выполнить задачу проекта и защитить персональные данные, методом их преобразования, необходимо будет:
1. Сгенерируем случайную квадратную матрицу по размеру количества признаков то есть 4х4.
2. Проверим сгениерированную матрицу на обратимость.
3. Умножить матрицу исходных признаков на новую созданную матрицу.
4. Определяем значения метрики качества R2 на исходных и преобразованных данных.
5. Делаем вывод о работе метода на основании сравнения метрики R2 для исходных и преобразованных данных.<br>
Таким образом исходные персональные данные поменяют свой вид и станут не читаемы.

**Обоснование**

Из предыдущего раздела мы выяснили что предсказания на исходной и измененной матрице не должны отличаться. Определить это можно по качеству обучения линейной регрессии. Определим метрику R2 на исходных и идменных данных.


## Проверка алгоритма

### Выделение признаков.

"Страховые выплаты" это целевой признак, остальные признаки.

In [5]:
features = data.drop('Страховые выплаты', axis=1)
target = data['Страховые выплаты']

### Создание модели линейной регрессии

In [6]:
class LinearRegression:
    def fit(self, train_features, train_target):
        X = np.concatenate((np.ones((train_features.shape[0], 1)), train_features), axis=1)
        y = train_target
        w = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)
        self.w = w[1:]
        self.w0 = w[0]

    def predict(self, test_features):
        return test_features.dot(self.w) + self.w0

### Создание обратимой матрицы

In [7]:
matrix = np.random.normal(size=(4, 4))
matrix

array([[-1.14287683, -0.10851307,  0.17552746, -1.35170531],
       [-0.87305587, -1.67171589,  0.09294288,  0.63680472],
       [ 0.44587228, -0.77937189, -0.09688664, -1.13705584],
       [-0.66520495, -0.01807249,  1.39367032,  1.12459473]])

Проверим сгенерированную матрицу на обратимость. Создадим обратную матрицу сгенерированной функцией np.linalg.inv(), если код выдаст ошибку, то сгенерированная матрица не обратимая.

In [8]:
matrix_inv = np.linalg.inv(matrix)
matrix_inv

array([[-0.52166672, -0.24890911,  0.60364912,  0.12426662],
       [ 0.15431227, -0.40605926, -0.43307083, -0.02246191],
       [ 0.00367519, -0.28982772,  0.60310691,  0.77832276],
       [-0.31064389,  0.20541642, -0.39730622, -0.00219502]])

Код ошибки не выдал, значит наша сгенерированная матрица обратима.

Получили квадратнцю матрицу размером 4х4

### Обучение модели на исходных данных

In [9]:
model_1 = LinearRegression()
model_1.fit(features, target)
predictions_1 = model_1.predict(features)
r2_1 = r2_score(target, predictions_1)
display(f'Значение метрики R2 до преборазования: {r2_1}')

'Значение метрики R2 до преборазования: 0.42494550286668'

### Обучение модели на преобразованных данных

Умножим признаки на сгенерированную матрицу.

In [10]:
new_features = features @ matrix
new_features

0             1            2             3
0     22077.661702 -38725.512799 -4800.197678 -56372.087910
1     16902.320854 -29693.048914 -3676.023423 -43177.704407
2      9337.999254 -16415.289501 -2031.924177 -23859.705358
3     18573.209481 -32534.950091 -4035.433907 -47399.606548
4     11611.678060 -20388.522949 -2525.963476 -29660.678665
...            ...           ...          ...           ...
4995  15891.864412 -27870.420748 -3453.463444 -40572.813859
4996  23333.358353 -40895.943573 -5072.306409 -59558.950197
4997  15096.278755 -26454.177614 -3279.811028 -38531.207780
4998  14557.677826 -25522.401361 -3161.791972 -37165.694270
4999  18076.160910 -31689.433461 -3929.426141 -46146.863788

[5000 rows x 4 columns]

Посчитаем значение метрики R2 после обучения модели на преобразованых данных.

In [11]:
model_2 = LinearRegression()
model_2.fit(new_features, target)
predictions_2 = model_2.predict(new_features)
r2_2 = r2_score(target, predictions_2)
display(f'Значение метрики R2 до преборазования: {r2_2}')

'Значение метрики R2 до преборазования: 0.42494550286668054'

**Вывод:**
Мы обучили модель линейной регрессии на исходных и преобразованных данных и посчитали значение метрики качаства R2. В обоих случаях значения метрик одинаковые. Это значит что использовать данные после их преобразования для предсказания страховых выплат будет корректно, без потери качества.<br>
А техника защиты данных методом умножения на обратимую матрицу работает и данные клиентов защищены.

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [ ]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования

## Финальные комментарии ревьюера

<div class="alert alert-success" style="border-color: green; border-radius: 5px">
    <p><u><b>✅ КОММЕНТАРИЙ РЕВЬЮЕРА</b></u></p>
    
Дополнительно рекомнедую почитать 

http://mathprofi.ru/deistviya_s_matricami.html здесь несколько разделов по линейной алгебре
    
https://www.youtube.com/watch?v=jUlO8aErxZM&list=PLcsjsqLLSfNAwTJdwpF6Qp2QHfqN62w33&index=1 а это, если надоели сериалы))
    
</div>